From 04c6a0771b62354e9b7be6f8865e4c9fa3c7519c Mon Sep 17 00:00:00 2001
From: Liang Qi <liang.qi@qt.io>
Date: Thu, 25 Sep 2025 13:25:04 +0200
Subject: [PATCH 213/553] cmake: adjust c++ standards for headers clean

Task-number: QTQAINFRA-7340
Pick-to: 6.8
Change-Id: Ib21a5a2670a67c8be2ba2606372bcdcc1c427676
Reviewed-by: Alexandru Croitor <alexandru.croitor@qt.io>
(cherry picked from commit 222663c18f05477837c92df6634de13a0c906a25)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 cmake/QtHeadersClean.cmake | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/cmake/QtHeadersClean.cmake b/cmake/QtHeadersClean.cmake
index 9df20aba946..f548a2321db 100644
--- a/cmake/QtHeadersClean.cmake
+++ b/cmake/QtHeadersClean.cmake
@@ -103,11 +103,19 @@ function(qt_internal_add_headersclean_target module_target module_headers)
     set(target_compile_flags_joined_genex
         "$<${compile_flags_exist_genex}:$<JOIN:${target_compile_flags_genex},;>>")
 
+    if (QT_FEATURE_cxx2c)
+        set(QT_HEADERS_CLEAN_CXX_STANDARD c++2c)
+    elseif (QT_FEATURE_cxx2b)
+        set(QT_HEADERS_CLEAN_CXX_STANDARD c++2b)
+    else()
+        set(QT_HEADERS_CLEAN_CXX_STANDARD c++20)
+    endif()
+
     if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU"
             OR "${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang|IntelLLVM")
 
-        # Compile header in strict C++20 mode. Enable further warnings.
-        set(hcleanFLAGS -std=c++2a
+        # Compile header in strict C++20 (or later) mode for C++17 build. Enable further warnings.
+        set(hcleanFLAGS -std=${QT_HEADERS_CLEAN_CXX_STANDARD}
             -Wall -Wextra -Werror -pedantic-errors
             -Woverloaded-virtual -Wshadow -Wundef -Wfloat-equal
             -Wnon-virtual-dtor -Wpointer-arith -Wformat-security
-- 
2.51.2

